src/Processors/QueryPlan/PartsSplitter.cpp

```diff
@@ -126,7 +126,9 @@ std::pair<std::vector<Values>, std::vector<RangesInDataParts>> split(RangesInDat
             return marks_in_current_layer < intersected_parts * 2;
         };
 
-        result_layers.emplace_back();
+        auto & current_layer = result_layers.emplace_back();
+        /// Map part_idx into index inside layer, used to merge marks from the same part into one reader
+        std::unordered_map<size_t, size_t> part_idx_in_layer;
```

Comment:
shouldn't it be just `ssize_t last_part_idx`?

Reply:
Looks like this will not be enough, since `parts_ranges_queue` is sorted by primary key value not by the part, plus there are modifications of this queue inside this loop.

Reply:
yep, I misunderstood your approach initially.